### ISSUE TITLE
Channel manager

### ISSUE THREAD

**Issue by towerofnix** [towerofnix/bantisocial#66]
Related to #14.

There needs to be a way to rename and delete channels (and probably reorder them) for admins.

## Todo

- [ ] Rename channels.
  - [ ] Backend. Probably: `POST /api/rename-channel {channelID, newName, sessionID}`.
  - [ ] Frontend. Either a literal "Rename" button, or maybe an edit (pencil) symbol beside the name.
- [ ] Delete channels?
  - This is complicated. We currently go out of our way to *not* delete messages - past revisions of edited messages are even stored and viewable. Following that, it doesn't make sense to delete an entire history of messages. Should we go with an "archive" option, instead? Archived channels are hidden from the obvious channel list, but they are not removed; anybody may view an archived channel, probably through a separate list.
- [ ] Reorder channels?
  - I'm for doing it The Discord Way, which means a global order that's unsortable to the client. But shrug. See discussion in #14.

**Comment by PullJosh** [comment]
> Either a literal "Rename" button, or maybe an edit (pencil) symbol beside the name.

Given that we have so many actions that we want users to be able to do (rename, delete, reorder, etc), I propose that each channel name has a three-dot context menu on the right side (only visible to admins). Something like this:

![image](https://user-images.githubusercontent.com/5458180/33629472-25f45f6e-d9d2-11e7-9daa-21b2417861e4.png)

(Clicking it, of course, would open a menu with a list of your possible actions.)

**Comment by PullJosh** [comment]
Re: Deleting channels

Personally, I think it makes the most sense to provide both "archive" and "delete" options. Archived channels could still be linked to and viewed, as well as re-activited by the admin if they want. However, they would be read-only and wouldn't appear in the regular channel list. This option would be best in the vast majority of cases.

However, I'm also in favor of giving as much power as possible to the users (particularly the trusted ones, like admins). Much like in the decision **not** to artificially limit the number of custom emoji that can be uploaded, I see no reason to prevent an admin from doing something (ie deleting a channel) if they think it's the best course of action. Admins are, presumably, trusted users, so there's no reason to believe that our own judgment as developers is any better than theirs. It's quite possible that some extenuating circumstances could arise that would make permanently deleting a channel the best possible option, so there's no reason to artificially prevent it.

**Comment by heyitsmeuralex** [comment]
>  I'm also in favor of giving as much power as possible to the users (particularly the trusted ones, like admins)

This is a good general design policy to go with - e.g. per-user channel order. :+1:

**Comment by towerofnix** [comment]
Opinions on de-archiving a channel? That seems reasonable to me, going with "let users do what they want".

**Comment by PullJosh** [comment]
I see no harm in letting it happen. Maybe leave a little note in the message list that says it was archived/unarchived?

**Comment by towerofnix** [comment]
Yup - I do want to figure out how to make "system" messages. Originally just for pinning messages ("blah pinned (a message)"), but also for this now.

**Comment by PullJosh** [comment]
Maybe when you click the "reorder" menu option, all of the channel names slide to the right a bit and a drag handle appears to the left of each? You could click and drag to your heart's content, and then click the "save" or "cancel" buttons which show up below the channel names.

Personally, I don't think the reorder option needs to specify that it's local or global. I think users would automatically assume that their changes only apply to themselves (or -- more likely -- wouldn't think about it at all).

If a regular user reorders their channels, it should save those settings for themselves only. If an admin reorders, the order that they choose should become the default for any new users who register in the future.

Every admin user should see the default layout at all times (rather than being allowed to set a custom order for themselves, like regular users can). So if one admin changes the order, that change should appear for all admins. Although this gives admins a bit less power, I think it simplifies their UI greatly, and avoids potential confusion between local and global order settings.

**Comment by heyitsmeuralex** [comment]
> So if one admin changes the order, that change should appear for all admins. Although this gives admins a bit less power

I disagree. I think that an admin changing channel order in that way should function exactly the same as a normal user doing the same. **Admins should have to go to the server settings to change the default channel order**. Similar drag+drog UI, but in a specific server settings dialog that explains that it's the _default_ for people who haven't changed the order yet/guests.

**Comment by PullJosh** [comment]
That probably makes more sense, actually.

**Comment by heyitsmeuralex** [comment]
That means this is blocked by #67, then.

**Comment by heyitsmeuralex** [comment]
Added `default channel option` to this massive issue as mentioned in #102.

**Comment by towerofnix** [comment]
Not blocked anymore since #67 is done; I'm also removing the "scary" label since #67 being implemented makes this take a lot less (client-side) work.
